packages/better-auth/src/api/middlewares/origin-check.ts

```diff
@@ -88,3 +81,134 @@ export const originCheck = (
 			validateURL(url, "callbackURL");
 		}
 	});
+
+/**
+ * Validates origin header against trusted origins.
+ * @param ctx - The endpoint context
+ * @param forceValidate - If true, always validate origin regardless of cookies/skip flags
+ */
+async function validateOrigin(
+	ctx: GenericEndpointContext,
+	forceValidate = false,
+): Promise<void> {
+	const headers = ctx.request?.headers;
+	if (!headers || !ctx.request) {
+		return;
+	}
+	const originHeader = headers.get("origin") || headers.get("referer") || "";
+	const useCookies = headers.has("cookie");
+
+	const shouldValidate =
+		forceValidate ||
+		(useCookies && !ctx.context.skipCSRFCheck && !ctx.context.skipOriginCheck);
+
+	if (!shouldValidate) {
+		return;
+	}
+
+	if (!originHeader || originHeader === "null") {
+		throw new APIError("FORBIDDEN", { message: "Missing or null Origin" });
+	}
+
+	// Handle dynamic trusted origins (function-based) or static (array-based)
+	const trustedOrigins: string[] = Array.isArray(
+		ctx.context.options.trustedOrigins,
+	)
+		? ctx.context.trustedOrigins
+		: [
+				...ctx.context.trustedOrigins,
+				...((await ctx.context.options.trustedOrigins?.(ctx.request)) || []),
+			];
+	const matchesPattern = (url: string, pattern: string): boolean => {
+		if (url.startsWith("/")) {
+			return false;
+		}
+		if (pattern.includes("*")) {
+			if (pattern.includes("://")) {
+				return wildcardMatch(pattern)(getOrigin(url) || url);
+			}
+			const host = getHost(url);
+			if (!host) {
+				return false;
+			}
+			return wildcardMatch(pattern)(host);
+		}
+		const protocol = getProtocol(url);
+		return protocol === "http:" || protocol === "https:" || !protocol
+			? pattern === getOrigin(url)
+			: url.startsWith(pattern);
+	};
+
+	const isTrustedOrigin = trustedOrigins.some((origin) =>
```

Comment:
I think you can now use the `context.isTrustedOrigin()` helper instead:

```suggestion
	const isTrustedOrigin = ctx.context.isTrustedOrigin(originHeader)
```